WalletWasabi.Tests/UnitTests/WabiSabi/Integration/Participant.cs

```diff
@@ -1,53 +1,46 @@
-using System.Collections.Generic;
-using System.Collections.Immutable;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using NBitcoin;
 using WalletWasabi.BitcoinCore.Rpc;
+using WalletWasabi.Blockchain.Analysis.Clustering;
 using WalletWasabi.Blockchain.Keys;
 using WalletWasabi.Blockchain.TransactionOutputs;
 using WalletWasabi.Blockchain.Transactions;
 using WalletWasabi.Models;
-using WalletWasabi.Tor.Socks5.Pool.Circuits;
 using WalletWasabi.WabiSabi.Client;
-using WalletWasabi.Wallets;
 using WalletWasabi.WebClients.Wasabi;
 
 namespace WalletWasabi.Tests.UnitTests.WabiSabi.Integration;
 
-internal class Participant
+internal class Participant: IDisposable
 {
-	public Participant(IRPCClient rpc, IWasabiHttpClientFactory httpClientFactory)
+	private bool _disposedValue;
+
+	public Participant(string name, IRPCClient rpc, IWasabiHttpClientFactory httpClientFactory)
 	{
-		Rpc = rpc;
 		HttpClientFactory = httpClientFactory;
 
-		KeyManager = KeyManager.CreateNew(out var _, password: "", Network.Main);
-		KeyManager.AssertCleanKeysIndexed();
+		Wallet = new TestWallet(name, rpc);
 	}
 
-	public KeyManager KeyManager { get; }
-	public List<SmartCoin> Coins { get; } = new();
-	public IRPCClient Rpc { get; }
+	private TestWallet Wallet { get; }
 	public IWasabiHttpClientFactory HttpClientFactory { get; }
-
-	private Coin? SourceCoin { get; set; }
+	private SmartTransaction? SplitTransaction { get; set; }
 
 	public async Task GenerateSourceCoinAsync(CancellationToken cancellationToken)
 	{
-		var minerKey = KeyManager.GetNextReceiveKey("coinbase", out _);
-		var blockIds = await Rpc.GenerateToAddressAsync(1, minerKey.GetP2wpkhAddress(Rpc.Network), cancellationToken).ConfigureAwait(false);
-		var block = await Rpc.GetBlockAsync(blockIds.First(), cancellationToken).ConfigureAwait(false);
-		SourceCoin = block.Transactions[0].Outputs.GetCoins(minerKey.P2wpkhScript).First();
-		minerKey.SetKeyState(KeyState.Used);
+		ThrowIfDisposed();
+		await Wallet.GenerateAsync(1, cancellationToken).ConfigureAwait(false);
 	}
 
 	public async Task GenerateCoinsAsync(int numberOfCoins, int seed, CancellationToken cancellationToken)
 	{
+		ThrowIfDisposed();
 		var feeRate = new FeeRate(4.0m);
-		var splitTx = Transaction.Create(Rpc.Network);
-		splitTx.Inputs.Add(new TxIn(SourceCoin!.Outpoint));
+		var splitTx = Wallet.CreateSelfTransfer(FeeRate.Zero);
+		var satoshisAvailable = splitTx.Outputs[0].Value.Satoshi;
+		splitTx.Outputs.Clear();
```

Comment:
How about having a `var spendableAmount = TestWallet.PrepareSpendingTx(FeeRate, out Transaction tx)`?

Reply:
ummm I don't understand it.

Reply:
The idea is the following, why create a transaction, get the available amount then remove the output to add it back later with a different value. It would be easier IMO to ask TestWallet for a "skeleton" tx and add the proper input outputs later.

Reply:
Right! it makes a lot of sense.